src/WebJobs.Script/Host/ScriptHost.cs

```diff
@@ -1089,11 +1089,13 @@ private void OnFileChanged(object sender, FileSystemEventArgs e)
             // - the host.json file was changed
             // - a function.json file was changed
             // - a function directory was added/removed
+            // - a function directory was renamed
             string fileName = Path.GetFileName(e.Name);
             if (isWatchedDirectory ||
                 ((string.Compare(fileName, ScriptConstants.HostMetadataFileName, StringComparison.OrdinalIgnoreCase) == 0) ||
                 string.Compare(fileName, ScriptConstants.FunctionMetadataFileName, StringComparison.OrdinalIgnoreCase) == 0) ||
-                (Directory.EnumerateDirectories(ScriptConfig.RootScriptPath).Count() != _directoryCountSnapshot))
+                (Directory.EnumerateDirectories(ScriptConfig.RootScriptPath).Count() != _directoryCountSnapshot) ||
+                (e.ChangeType == WatcherChangeTypes.Renamed && String.IsNullOrEmpty(Path.GetExtension(e.FullPath))))
```

Comment:
This will trip with extensionless files and directory renames inside of function directories, which shouldn't require a restart. 
You'd likely need to check the file attributes to ensure it's a directory and only restart if if is a directory in the script root (a function directory)
With those requirements, it may be more efficient (given the file system constraints we have) and simpler to replace the `_directoryCountSnapshot` with the actual result of `Directory.EnumerateDirectories` when we assign it (changing it to an `IEnumerable<string>`) and change the logic above to use the results of `EnumerateDirectories` to compare against your snapshot, a quick `Count` check and zipping through the results would be sufficient to tell you if you need to restart.